### ISSUE TITLE
vim9: Unexpected: E691: Can only compare List with List

### ISSUE THREAD

**Issue by errael** [vim/vim#10022]
### Steps to reproduce

In the following, the identified line gets the error.
```
vim9script

# Find target in nested list using '=='
def Find(target: any, l: list<any>)
    def FindInternal(lin: list<any>)
        var i = 0
        while i < len(lin)
            var this_one: any = lin[i]
            if this_one == target          # <<<<<<<<<<<<<<<<<<<
                echo 'found: ' this_one
            elseif type(this_one) == v:t_list
                FindInternal(this_one)
            endif
            i += 1
        endwhile
    enddef
    FindInternal(l)
enddef
Find(1, [0, 1, [ 0, 1, 2], 1])
```
Initially had `var this_one = lin[i]` (the line above the identified line).

After seeing the error, I tried `var this_one: any = lin[i]` to no avail.

Seems like comparing `any` to `any` should not be an error.

### Expected behaviour

Comparing `any` to `any` should not be an error.

Following is a workaround
```
if type(this_one) == type(target) && this_one == target
```

### Version of Vim

8.2.4628

### Environment

ubuntu, gvim gtk3

**Comment by brammool** [comment]
Note that the error starts with "Error detected while processing", which means it happens when the code is executed, not when it is compiled.
If you add this above the line that produces the error:
```
echomsg typename(this_one) 'equals' typename(target)
```
you will see "list<number> equals number", and that is what is causing the error.

The "any" type can be useful when compiling and you don't know the type yet.
But at runtime the types do need to match for "==" to work.

**Comment by errael** [comment]
> it happens when the code is executed

Right, that's why i added `type(this_one) == type(target) &&` before the _real_ compare as a workaround.

I was working on something to search an arbitrary structure. Other than that, I've not seen this error; and I agree I shouldn't ever see it doing regular things.

I'm not familiar enough with modern languages like Kotlin/scala, but I'm surprised that in strict typing language, something that compiles would have runtime error about an equality test.
 
I added EQ() and IS() to my toolkit, but shouldn't use them often. My only concern is getting a run time error in the field when the code is "correct". Guess I'll have to pay attention to what I'm doing.

Closing.
